File: java/2020/day4/day4.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class Day4 {

    public static boolean isValid(String passport) {
        String[] checks = {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};
        for (String key : checks) {
            if (passport.indexOf(key) == -1) {
                return false;
            }
        }
        return true;
    }
    public static void main(String[] args) throws FileNotFoundException {
        Scanner file = new Scanner(new File("./day4.in"));

        ArrayList<Passport> passports = new ArrayList<>();

        String currentPass = "";
        while (file.hasNextLine()) {
            String line = file.nextLine();
            if (line.length() != 0) {
                currentPass += line.strip() + " ";
            } else {
                passports.add(new Passport(currentPass.strip()));
                currentPass = "";
            }
        }
        passports.add(new Passport(currentPass.strip())); // adds last passport

        int p1Valid = 0;
        int p2Valid = 0;

        for (Passport passport : passports) {
            if (passport.containsRequiredKeys()) {
                p1Valid++;
            }
            if (passport.isValid()) {
                p2Valid++;
            }
        }

        System.out.println(p1Valid);
        System.out.println(p2Valid);
        
        file.close();
    }
}

class Passport {
    
    private int birthYear, issueYear, expirationYear;
    private String height, hairColor, eyeColor, passID;
    private String passportString;
    private boolean valid = true;

    public static boolean containsRequiredKeys(String passport) {
        String[] checks = {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};
        for (String key : checks) {
            if (passport.indexOf(key) == -1) {
                return false;
            }
        }
        return true;
    }

    public boolean containsRequiredKeys() {
        return containsRequiredKeys(passportString);
    }

    private static String[] sortAlphabetically(String[] strings) {
        String temp;
        for (int i = 0; i < strings.length; i++) 
        {
            for (int j = i + 1; j < strings.length; j++) { 
                if (strings[i].compareTo(strings[j])>0) 
                {
                    temp = strings[i];
                    strings[i] = strings[j];
                    strings[j] = temp;
                }
            }
        }
        return strings;
    }

    private static String[] removeCID(String[] passport) {
        String[] result = new String[7];
        int index = 0;
        for (String key : passport) {
            if (!key.startsWith("cid")) {
                result[index] = key;
                index++;
            }
        }
        return result;
    }

    private static String[] createPassportArray(String passport) {
        String[] result = sortAlphabetically(passport.split(" "));
        // when sorted, key order is "byr", "cid", "ecl", "eyr", "hcl", "hgt", "iyr", "pid"
        if (result.length == 8) {
            result = removeCID(result);
        }
        for (int i = 0; i < result.length; i++) {
            result[i] = result[i].substring(4);
        }
        return result;
        
    }

    public Passport(String passport) {
        passportString = passport;
        if (containsRequiredKeys(passport)) {
            String[] pass = createPassportArray(passport);
            // key order is "byr" (0), "ecl" (1), "eyr" (2), "hcl" (3), "hgt" (4), "iyr" (5), "pid" (6)
            birthYear = Integer.parseInt(pass[0]);
            eyeColor = pass[1];
            expirationYear = Integer.parseInt(pass[2]);
            hairColor = pass[3];
            height = pass[4];
            issueYear = Integer.parseInt(pass[5]);
            passID = pass[6];
        } else {
            valid = false;
        }
    }


    public boolean validBirthYear() {
        return (birthYear >= 1920 && birthYear <= 2002);
    }

    public boolean validEyeColor() {
        String[] validColors = {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"};
        for (String c : validColors) {
            if (eyeColor.equals(c)) {
                return true;
            }
        }
        return false;
    }

    public boolean validExpirationYear() {
        return (expirationYear >= 2020 && expirationYear <= 2030);
    }

    public boolean validHairColor() {
        if (hairColor.length() == 7 && hairColor.charAt(0) == '#') {
            for (int i = 1; i < hairColor.length(); i++) {
                if ("0123456789abcdef".indexOf(hairColor.charAt(i)) == -1) {
                    return false;
                }
            }
            return true;
        }
        //System.out.println(hairColor);
        return false;
    }

    public boolean validHeight() {
        if (height.endsWith("cm")) {
            int val = Integer.parseInt(height.substring(0, height.length()-2));
            if (val >= 150 && val <= 193) {
                return true;
            }
        } else if (height.endsWith("in")) {
            int val = Integer.parseInt(height.substring(0, height.length()-2));
            if (val >= 59 && val <= 76) {
                return true;
            }
        }
        return false;
    }

    public boolean validIssueYear() {
        return (issueYear >= 2010 && issueYear <= 2020);
    }

    public boolean validPassportID() {
        try {
            Integer.parseInt(passID);
            
            if (passID.length() == 9) {
                return true;
            }
        } catch (NumberFormatException e) {
            System.out.println(passID);
            return false;
        } 
        return false;
    }

    public boolean isValid() {
        return (valid && validBirthYear() && validExpirationYear() && validEyeColor() && validHairColor() && validHeight() && validIssueYear() && validPassportID());
    }

}

File: cpp/2020/day4/day4.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>

using namespace std;

template <class T>
class undefinable {
    private:
        T val;
        bool defined;

    public:
        undefinable() {
            defined = false;
        }
        T operator= (T value) {
            val = value;
            defined = true;
            return value;
        }
        bool is_defined() {
            return defined;
        }
        T value() {
            return val;
        }
};

struct passport {
    undefinable<unsigned int> birth_yr, issue_yr, expire_yr; 
    undefinable<string> height, hair_clr, eye_clr, pid;

    string in;

    passport(const string &input) {
        string field, key, value;
        stringstream stream;
        stream << input;

        in = input;

        while (stream >> field) {
            key = field.substr(0, 3);
            value = field.substr(4);

            if (key == "byr") {
                birth_yr = stoi(value);
            } else if (key == "iyr") {
                issue_yr = stoi(value);
            } else if (key == "eyr") {
                expire_yr = stoi(value);
            } else if (key == "hgt") {
                height = value;
            } else if (key == "hcl") {
                hair_clr = value;
            } else if (key == "ecl") {
                eye_clr = value;
            } else if (key == "pid") {
                pid = value;
            }
        }
    } 

    bool all_assigned() {
        return (birth_yr.is_defined() && issue_yr.is_defined() && expire_yr.is_defined() && height.is_defined() && 
            hair_clr.is_defined() && eye_clr.is_defined() && pid.is_defined());
    }

    bool byr_valid() {
        return (birth_yr.value() >= 1920 && birth_yr.value() <= 2002);
    }

    bool iyr_valid() {
        return (issue_yr.value() >= 2010 && issue_yr.value() <= 2020);
    }

    bool eyr_valid() {
        return (expire_yr.value() >= 2020 && expire_yr.value() <= 2030);
    }

    bool hgt_valid() {
        const string &h = height.value();
        if (h.length() < 2) return false;

        string end = h.substr(h.length() - 2);  
        unsigned int val;
        try {
            val = stoi(h.substr(0, h.length() - 2));
        } catch (const std::exception &e) {
            return false;
        }


        return ((end == "cm" && val >= 150 && val <= 193) || (end == "in" && val >= 59 && val <= 76));
    }

    bool hcl_valid() {
        if (hair_clr.value()[0] == '#' && hair_clr.value().length() == 7) {
            for (char c : hair_clr.value().substr(1)) {
                if ((c < '0' || c > '9') && (c < 'a' || c > 'f')) {
                    cout << "char:" << c << endl;
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    bool ecl_valid() {
        string colors[] = {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"};

        for (string c : colors) if (eye_clr.value() == c) return true;
        return false;
    }

    bool pid_valid() {
        if (pid.value().length() == 9) {
            for (char digit : pid.value()) {
                if (digit < '0' || digit > '9') {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    bool all_valid() {
        return (all_assigned() && byr_valid() && eyr_valid() && iyr_valid() && hcl_valid() && ecl_valid() && pid_valid() && hgt_valid());
    }
};

void insert_data(vector<passport> &data) {
    ifstream infile;
    infile.open("./day4.in");

    string line, pp_string = "";

    while (getline(infile, line)) {
        if (line != "") {
            pp_string += " " + line;
        } else {
            passport pp(pp_string);
            data.push_back(pp);
            pp_string = "";
        }
    }
    passport pp(pp_string);
    data.push_back(pp);

    infile.close();
}

int main() {

    vector<passport> data;
    insert_data(data);

    int p1_count = 0, p2_count = 0;

    for (passport p : data) {
        if (p.all_assigned()) {
            p1_count++;
        }
        if (p.all_valid()) {
            p2_count++;
        }
    }

    cout << p1_count << "\n" << p2_count << "\n";

    return 0;
}

